feat: Background submissions for submittable doctypes

Aradhya-Tripathi opened the pull request:
Option to configure background submissions for submittable doctypes

#

TODO:
- [x] locking in framework is messed up - need to fix this first
- [x] Make notification redirect to the reference document
- [x] provide a realtime alert if the user is logged in (?)
- [x] Execute other `savedocs` stuff (?)
- [x] Unlock doc custom button (from frontend - only if the job doesn't exist in queue)
- [x] Add `queue_in_background` to customize form as well
- [x] Validation for setting `queue_in_background` to `0` if doctype is not-submittable
- [x] add `Submission Queue` to log settings - for regular cleanup (30 days)
- [x] UI for showing previous failed submission
- [x] if user is on the same form while the submit completes - it should be reloaded
- [x] add queueing in bulk submit
- [x] Test with erpnext (along with server scripts)
- [x] Tests

Commit:
feat: Background submissions for submittable doctypes

Commit:
Merge branch 'bg-submissions' of https://github.com/Aradhya-Tripathi/frappe into bg-submissions

ankush commented on frappe/desk/form/save.py:
```diff
@@ -17,9 +17,15 @@ def savedocs(doc, action):
 
 	if doc.docstatus == 1:
-		doc.submit()
+		if doc.bg_submit and doc.is_submittable:
```
These properties are of `doc.meta` and not document itself. 

doc.meta contains the metaclass for a document

phot0n replied:
I'm kinda thinking we should not give a user facing option for this - could lead to a lot of delays (unnecessary queuing of things even if they're not required which could block the queue for the jobs that *really* require queueing) (?) 🤔 

probably a (bool) class variable on the doctype controller (like `queue_on_submit` ?)

ankush replied:
Nope. This should be user configurable 😅  

Not every site needs this, not every doctype needs this by default. Just on some sites where volumes are high enough and data is large enough this will help.

Commit:
refactor: changed field name & removed queueing from save

phot0n commented on frappe/desk/form/save.py:
```diff
@@ -16,9 +16,11 @@ def savedocs(doc, action):
-
 	if doc.docstatus == 1:
-		doc.submit()
+		if doc.meta.submit_in_background and doc.meta.is_submittable:
```
```suggestion
		if doc.meta.submit_in_background:
```

#

Probably no need to check issubmittable ...If docstatus is 1 it'll be 1 of 2 cases:

a) the document is being submitted
b) the submitted document is being updated

phot0n replied:
also i'm kinda wondering why you're doing this in `savedocs` function? 

this would probably be a better place?

https://github.com/frappe/frappe/blob/77be534e225aebaaa88828b08b067d1d2ba52d5b/frappe/model/document.py#L1010-L1012

since this is what's being called (mostly) for submitting a document everywhere (if not overridden)

phot0n replied:
A problem that could arise from this is, if someone has done

```
x = doc.submit()
```

We return the whole doc object (i think) after submission and in case of queueing we won't get that and it will cause a lot of problems :P

Aradhya-Tripathi replied:
hey but I don't think there would be immediate breaking changes, since by default it's 0

ankush replied:
Submitting from code and submitting from frontend is different right? 

We can expose same API for when someone wants to submit things from code. Doing this in code wouln't make sense.


Imagine a code that expects a submitted document after calling `doc.submit` they wouldn't actually see a submitted document.

phot0n replied:
let's do this via savedocs ..seems waayyy safer

Aradhya-Tripathi committed:
refactor: removed excess checks

Co-authored-by: Ritwik Puri <ritwikpuri5678@gmail.com>

Commit:
Merge branch 'develop' of https://github.com/frappe/frappe into bg-submissions

Commit:
feat: log reports for queued submits

phot0n commented on frappe/core/doctype/queued_submit/queued_submit.json:
```diff
@@ -0,0 +1,80 @@
+ "modified": "2022-10-03 17:28:12.081732",
+ "modified_by": "Administrator",
+ "module": "Core",
+ "name": "Queued Submit",
```
maybe `submission queue`?

Also, use hash based naming for this (or maybe autoincrement :P)

phot0n commented on frappe/core/doctype/queued_submit/queued_submit.json:
```diff
@@ -0,0 +1,80 @@
+   "role": "System Manager",
+   "share": 1,
+   "write": 1
+  }
```
```suggestion
   "create": 0,
   "delete": 0,
   "email": 0,
   "export": 0,
   "print": 0,
   "read": 1,
   "report": 0,
   "role": "System Manager",
   "share": 0,
   "write": 0
  }
```

permissions to the doctype should be managed ..only allow reads from frontend (writing/updation will be done via backend)

phot0n commented on frappe/core/doctype/queued_submit/queued_submit.json:
```diff
@@ -0,0 +1,80 @@
+  "state",
+  "start_time",
+  "created_by",
+  "error"
```
maybe something more generic? like `message`?

phot0n commented on frappe/core/doctype/queued_submit/queued_submit.json:
```diff
@@ -0,0 +1,80 @@
+   "fieldname": "state",
+   "fieldtype": "Data",
+   "label": "State"
+  },
```
this will be a `Select` field ..with Queued, Completed, Failed as it's options

phot0n commented on frappe/core/doctype/queued_submit/queued_submit.json:
```diff
@@ -0,0 +1,80 @@
+ "editable_grid": 1,
+ "engine": "InnoDB",
+ "field_order": [
+  "title",
```
what's this field used for?

phot0n commented on frappe/model/document.py:
```diff
@@ -984,6 +985,28 @@ def _evaluate_alert(alert):
+			doc.state = "Failed"
+			doc.error = str(e)
+			doc.insert()
+
```
both these methods should not be present in `document.py` ..doesn't make sense

phot0n commented on frappe/core/doctype/queued_submit/queued_submit.json:
```diff
@@ -0,0 +1,80 @@
+ "doctype": "DocType",
+ "editable_grid": 1,
+ "engine": "InnoDB",
+ "field_order": [
```
reference doctype and docname fields should also be added :P

(dynamic link)

phot0n commented on frappe/model/document.py:
```diff
@@ -984,6 +985,28 @@ def _evaluate_alert(alert):
+			doc = frappe.get_doc("Queued Submit", self.title)
+			doc.state = "Failed"
+			doc.error = str(e)
+			doc.insert()
```
```suggestion
			frappe.db.set_value("Queued Submit", self.title, {"state": "Failed", "error": str(e)})
```

probably don't need the whole doc object here

phot0n commented on frappe/model/document.py:
```diff
@@ -984,6 +985,28 @@ def _evaluate_alert(alert):
+			self.submit()
+			doc = frappe.get_doc("Queued Submit", self.title)
+			doc.state = "Submitted"
+			doc.insert()
```
```suggestion
			frappe.db.set_value("Queued Submit", self.title, {"state": "Completed"})
```

probably don't need the whole doc object here

phot0n commented on frappe/core/doctype/doctype/doctype.json:
```diff
@@ -605,6 +606,12 @@
    "fieldname": "make_attachments_public",
    "fieldtype": "Check",
    "label": "Make Attachments Public by Default"
+  },
+  {
+   "default": "0",
+   "fieldname": "submit_in_background",
+   "fieldtype": "Check",
+   "label": "Submit in background"
```
should only be shown when `issubmittable` is checked

the `depends_on` property can be used for this - check other examples in this file

phot0n replied:
also this does not seem to be easily accessible.

would be better to show over here

<img width="1083" alt="Screenshot 2022-10-05 at 3 14 41 PM" src="https://user-images.githubusercontent.com/32034600/194031618-17aee9ec-146d-40f1-8a16-5a95b37d97e0.png">

Aradhya-Tripathi replied:
done 👍🏿

Commit:
fix: fixed functionality

phot0n commented:
somehow the doctype got deleted in the recent commit 🤣

Commit:
refactor: moved submission from document class

Commit:
fix: unlock document when action is executed

Commit:
refactor: moved background submit to Submission Queue DocType

Co-authored-by: phot0n ritwikpuri5678@gmail.com

phot0n commented on frappe/core/doctype/submission_queue/submission_queue.py:
```diff
@@ -0,0 +1,67 @@
+			"type": "Notification",
+			"document_type": "Submission Queue",
+			"document_name": name,
+			"subject": "Job Queued",
```
subject should be something different? at this point the job has either failed or succeeded

phot0n commented on frappe/model/document.py:
```diff
@@ -3,6 +3,7 @@
 import hashlib
 import json
 import time
+from datetime import datetime
```
revert?

Aradhya-Tripathi replied:
👍🏿

phot0n commented on frappe/core/doctype/submission_queue/submission_queue.py:
```diff
@@ -0,0 +1,67 @@
+			action_for_queuing=self.action_for_queuing,
+		)
+		frappe.db.set_value(self.doctype, self.name, {"job_id": job.id}, update_modified=False)
+
```
```suggestion
```

phot0n commented on frappe/core/doctype/submission_queue/submission_queue.py:
```diff
@@ -0,0 +1,67 @@
+		"Submission Queue", filters=name, fieldname="created_by"
+	), fieldname="email")
+	if mention:
+		enqueue_create_notification([mention], notification_doc)
```
we can pass the `created_by` value of submission queue doc directly to this function? no need to fetch that again (need to add a param for this tho)

phot0n commented on frappe/core/doctype/submission_queue/submission_queue.py:
```diff
@@ -0,0 +1,67 @@
+			getattr(to_be_queued_doc, _action)()
+			values = {"status": "Completed"}
+		except Exception as e:
+			values = {"status": "Failed", "message": str(e)}
```
this doesn't log the whole traceback ..only the exception message. we can use `frappe.get_traceback` for this.

phot0n commented on frappe/core/doctype/submission_queue/submission_queue.py:
```diff
@@ -0,0 +1,67 @@
+
+def notify(name: str):
+	notification_doc =  {
+			"type": "Notification",
```
type should be `Alert`?

there is no option of Notification type in Notification Log

https://github.com/frappe/frappe/blob/bd21398252a4b84e8044b7d5a60b2a89f6223ede/frappe/desk/doctype/notification_log/notification_log.json#L35-L41

phot0n commented on frappe/core/doctype/submission_queue/submission_queue.py:
```diff
@@ -0,0 +1,67 @@
+	notification_doc =  {
+			"type": "Notification",
+			"document_type": "Submission Queue",
+			"document_name": name,
```
we should take the user to the submission queue only when the job has failed.. otherwise we can log the actual doc itself? we can use the status key in values dict for this

phot0n replied:
this is something we need to figure out ..upon clicking the notification we should be redirected to the referenced document ..but somehow we aren't

phot0n commented on frappe/core/doctype/submission_queue/submission_queue.py:
```diff
@@ -0,0 +1,67 @@
+		super().insert()
+
+	def after_insert(self):
+		job = self.queue_action(
+			"queue_in_background",
+			to_be_queued_doc=self.to_be_queued_doc,
+			action_for_queuing=self.action_for_queuing,
+		)
```
also this is wrong :sweat_smile: - my bad. this will lock the submission queue doc rather than the `to_be_queued_doc` ..probably the reason why I don't get any error upon doing multiple submissions.

let me know if you need help with this.

Aradhya-Tripathi replied:
fixed

phot0n commented on frappe/core/doctype/submission_queue/submission_queue.json:
```diff
@@ -0,0 +1,112 @@
+ "engine": "InnoDB",
+ "field_order": [
+  "status",
+  "start_time",
+  "created_by",
```
```suggestion
  "enqueued_at",
  "enqueued_by",
```

much better naming ;P

Aradhya-Tripathi replied:
done

phot0n commented on frappe/core/doctype/submission_queue/submission_queue.py:
```diff
@@ -0,0 +1,67 @@
+			"document_name": name,
+			"subject": "Job Queued",
+			"from_user": frappe.session.user,
+			"email_content": "Hello",
```
we can remove this i think (?)

phot0n commented on frappe/core/doctype/submission_queue/submission_queue.json:
```diff
@@ -0,0 +1,112 @@
+  "ref_docname",
+  "section_break_8",
+  "message"
+ ],
```
also we should probably add a `completed_at` field

Aradhya-Tripathi replied:
added

Commit:
fix: locking the correct doctype

Commit:
refactor: better locking and unlocking

Commit:
feat: Added completed_at field and better naming

Commit:
refactor: changed field order & added depends on

phot0n committed:
refactor(minor): better notification

* changed completed_at fieldname to ended_at (in submission queue doctype)
* added rollback on exception in queue_in_background method

phot0n committed:
refactor(minor): use virtual docfield for created_at (previously enqueued_at) field

phot0n commented on frappe/core/doctype/doctype/doctype.json:
```diff
@@ -605,6 +606,13 @@
+  },
+  {
+   "default": "0",
+   "depends_on": "is_submittable",
```
```suggestion
   "depends_on": "eval:doc.is_submittable",
```

Aradhya-Tripathi replied:
fixed

Commit:
fix: only show submit in background when doc is submittable

* chore: better naming

phot0n committed:
fix(minor): show ref_doctype and ref_docname in listview of submission queue

phot0n committed:
fix(minor): only queue in background for submit action

Commit:
feat: Added unlocked button for locked documents not in queue

Commit:
feat: Added status updates in documents

phot0n commented on frappe/core/doctype/submission_queue/submission_queue.py:
```diff
@@ -0,0 +1,118 @@
+
+	@frappe.whitelist()
+	def unlock_doc(self):
+		if self.is_locked:
+			try:
+				Job(self.job_id, connection=get_redis_conn())
+				frappe.msgprint(_("Document already exists in queue!"))
+			except NoSuchJobError:
+				self.to_be_queued_doc.unlock()
+				self.status = "Failed"
+				self.save()
+				frappe.msgprint(_("Unlocked document as no such document exists in queue"))
+		else:
+			# failed, completed don't know at this point
+			self.status = "Failed"
+			self.save()
+			frappe.msgprint(_("Document is already unlocked"))
```
```suggestion
		try:
			Job.fetch(self.job_id, connection=get_redis_conn())
			frappe.msgprint(_("Job for submission of the reference document exists in queue"), _("Failed to unlock"))
		except NoSuchJobError:
			self.unlock()
			frappe.msgprint(_("Reference Document Unlocked"))
```

TBH not really sure if the status can be considered `Failed` in the except block, since the job can also be killed/timed out after the whole submission scenario happens ..best to leave it in the `Queued` state?

phot0n replied:
Plus i'm kinda unsure about the `Job.fetch` method - able to fetch the job even after it's completed/doesn't exist in the queue

<img width="1185" alt="Screenshot 2022-10-08 at 3 22 40 AM" src="https://user-images.githubusercontent.com/32034600/194668046-6acada2c-4e2b-4387-a0c4-40aaa5f5b4cc.png">

#

maybe we can use this?
https://github.com/rq/rq/blob/4750eb43161cf6d2dee5cab4c20b7257476fbc1b/rq/queue.py#L215-L218

Though this will become a O(n) operation rather than O(1) - maybe something better exists 🤔

Aradhya-Tripathi replied:
`job.heart_beat` would be better

Aradhya-Tripathi replied:
We should also have a better way of enforcing job status for documents which are not in rq's queue but are marked as Queued

phot0n commented on frappe/model/document.py:
```diff
@@ -1492,12 +1493,14 @@ def lock(self, timeout=None):
 		if self in frappe.local.locked_documents:
 			frappe.local.locked_documents.remove(self)
+			self.is_locked = False
 
```
revert?

Aradhya-Tripathi replied:
I feel like a doc property should exist telling the status of the doc? as we don't have another docstatus for locked state🤔

phot0n commented on frappe/core/doctype/submission_queue/submission_queue.js:
```diff
@@ -0,0 +1,10 @@
+
+frappe.ui.form.on('Submission Queue', {
+    refresh: function(frm) {
+      frm.add_custom_button(__("Unlock"), () => {
+		frm.call("unlock_doc")
+	  })
```
```suggestion
      if (frm.doc.status === "Queued") {
          frm.add_custom_button(__("Unlock Reference Document"), () => {
		    frm.call("unlock_doc")
	      });
	  }
```

only show the button when the status is in `Queued` state?

Aradhya-Tripathi replied:
👍🏿

Commit:
fix: fixed job status after fetching job

Commit:
fix: checking and unlocking the correct doc

Commit:
feat: Added more statuses for document state

Commit:
fix: checking if doc is submittable before queueing

Commit:
fix(minor): better naming

phot0n commented on frappe/desk/form/save.py:
```diff
@@ -16,8 +18,15 @@ def savedocs(doc, action):
+		if (
+			action == "Submit"
+			and doc.meta.queue_in_background
+			and doc.meta.is_submittable
```
```suggestion
```

while this works this makes it hard for any other person looking into this. they will easily mistake this as being an additional unnecessary check and remove it, which will cause problems. 

please make sure the validation exist in `doctype.py` as well as `customize_form.py`.

Commit:
refactor: added more states to document when it's queued

Commit:
fix: fixed unlocking of docs from frontend

Aradhya-Tripathi commented:
Need to fix notifications to allow redirect

Commit:
refactor: removed redundant checking of locked documents

Commit:
feat: Adding queue in background to customize form as well

Commit:
fix: fixed redirection to doctype on notification

Commit:
fix: temporary document locking

Commit:
feat: Adding data to monitor on action

Commit:
feat: Added identifier for locked state of documents

refactor(minor): removed is locked setter to avoid redundancy

phot0n commented on frappe/model/document.py:
```diff
@@ -1412,7 +1415,9 @@ def log_error(self, title=None, message=None):
 		"""Returns signature (hash) for private URL."""
-		return hashlib.sha224(get_datetime_str(self.creation).encode()).hexdigest()
+		return hashlib.sha224(
+			get_datetime_str(self.creation or self.modified or now()).encode()
```
this is bad ..`self.modified` keeps on changing via db api's and other places as well. `creation` is much safer in that context

since the signature is used to unlock the doc it needs to be the same for that particular cycle of locking/unlocking.

phot0n commented on frappe/core/doctype/submission_queue/submission_queue.py:
```diff
@@ -0,0 +1,135 @@
+
+	def unlock_doc_and_update_status(self, doc_to_be_unlocked: Document, possible_status: tuple):
+		try:
+			job = Job(self.job_id, connection=get_redis_conn())
```
this should be `Job.fetch`?

phot0n commented on frappe/core/doctype/submission_queue/submission_queue.py:
```diff
@@ -0,0 +1,135 @@
+
+	def notify(self, submission_status: str, action: str):
+		if submission_status == "Failed":
+			doctype = "Submission Queue"
```
```suggestion
			doctype = self.doctype
```

phot0n commented on frappe/core/doctype/submission_queue/submission_queue.py:
```diff
@@ -0,0 +1,135 @@
+		except NoSuchJobError:
+			# Need to update status
+			doc_to_be_unlocked.unlock()
+			frappe.msgprint(_("Unlocked document as no such document exists in queue"))
```
```suggestion
			frappe.msgprint(_("Document Unlocked"))
```

phot0n commented on frappe/core/doctype/submission_queue/submission_queue.py:
```diff
@@ -0,0 +1,135 @@
+			if status in possible_status:
+				doc_to_be_unlocked.unlock()
+				self.status = status.capitalize()
+				self.save()
```
directly use `db.set_value`? - `self.save` is too many checks for just updating status of the submission queue

phot0n commented on frappe/core/doctype/submission_queue/submission_queue.py:
```diff
@@ -0,0 +1,135 @@
+			message = _("Submission of {0} {1} with action {2} completed successfully")
+
+		notification_doc = {
+			"type": "Mention",
```
raised a fix for this -> #18334 

can be reverted back to alert :)

Commit:
feat: Adding method to clear logs

phot0n commented on frappe/model/document.py:
```diff
@@ -309,6 +317,8 @@ def _save(self, ignore_permissions=None, ignore_version=None) -> "Document":
 			return
 
+		self.check_locked_document()
+
```
this won't matter in case of insertion of a document ..since it doesn;t really exist at that point

might wanna move this before `self.check_permission("write", "save")` (line 322/332)

phot0n commented on frappe/model/document.py:
```diff
@@ -987,12 +997,14 @@ def _evaluate_alert(alert):
 	@whitelist.__func__
 	def _cancel(self):
 		"""Cancel the document. Sets `docstatus` = 2, then saves."""
+		self.check_locked_document()
```
revert?

phot0n commented on frappe/model/document.py:
```diff
@@ -987,12 +997,14 @@ def _evaluate_alert(alert):
 	@whitelist.__func__
 	def _submit(self):
 		"""Submit the document. Sets `docstatus` = 1, then saves."""
+		self.check_locked_document()
```
revert?

phot0n commented on frappe/core/doctype/submission_queue/submission_queue.py:
```diff
@@ -0,0 +1,143 @@
+
+	@frappe.whitelist()
+	def unlock_doc(self):
+		possible_status = ("failed", "canceled", "stopped", "finished")
```
better naming?

Commit:
refactor: edited unlocked document message

Commit:
refactor: moved is_submittable validation from save to submission queue

Commit:
fix: fixed locking and checking

Commit:
feat: Added submittable and queue in background logic

Commit:
refactor: better naming

Commit:
feat: Added auto log clearing

Commit:
Merge branch 'develop' of https://github.com/frappe/frappe into bg-submissions

Commit:
feat: introduced file for tracking queues

phot0n committed:
refactor(minor): made unlock_doc_and_update_status into a simple function

* renamed unlock_doc_and_update_status -> unlock_reference_doc
* added queued_doc property
* renamed check_locked_document -> check_if_locked
* reduced the statuses in submission queue
* refactored unlock_reference_doc a bit

Commit:
feat: realtime alerts if user is logged in

phot0n committed:
chore: remove unnecessary condition

Commit:
feat: Redirecting to doctype on alert

Commit:
Merge branch 'bg-submissions' of https://github.com/Aradhya-Tripathi/frappe into bg-submissions

Commit:
feat: clearning submission queue entries & additional check for unlocking documents

Commit:
refactor: removed state management via json file for queues

Commit:
fix: added ignore permissions flag for doctype creations by other users

Commit:
test: checking persistant locks on document instances

Commit:
refactor: removed unnecesary actions from queue

Commit:
fix: fixed realtime updates

Commit:
feat: Added on_success and on_failure to enqueue

Commit:
test: Added test for asserting queue action

Commit:
test: fixed queueing test and added completion check

Commit:
test: removed completion check

Commit:
fix: merge conflicts

netchampfaris commented:
@Aradhya-Tripathi Is this only for the `submit` action? I was hoping this would be generic for any long-running action on a document.

Aradhya-Tripathi commented:
> @Aradhya-Tripathi Is this only for the `submit` action? I was hoping this would be generic for any long-running action on a document.

For starters I made it only for submittable doctypes seeing submittable doctypes in erpnext how ever planning on extending to other long running actions as well.

phot0n committed:
chore: rename queue to background_submission

phot0n committed:
fix: realtime alert

use session object's last updated time to figure out whether to send a realtime alert or a notification

phot0n committed:
refactor(minor): move staticmethod after properties

phot0n committed:
refactor(minor): remove additional document states and rename message to exception

phot0n committed:
feat(minor): submission queue banner for showing recent submissions

phot0n commented on frappe/core/doctype/submission_queue/submission_queue.py:
```diff
@@ -0,0 +1,187 @@
+			# assuming the job failed here (?)
+			status = "failed"
+
+		# Job finished successfully however action was never completed (?)
+		if status == "finished" and self.queued_doc.docstatus != 1:
+			status = "failed"
```
when will this be true?

if the job finishes i think in this case it will most likely update the submission queue status to `Failed`? - anything i'm missing :thinking:

Aradhya-Tripathi replied:
Yes, so sometimes it doesn't reach that part, locally I have a few docs whose execution is finished according to rq without error however their docstatus remains as draft

phot0n replied:
weird that this is happening - ideally if a job has "finished" it should either mean the doc failed/passed the submission :thinking: 

maybe there's something we're missing (?)

phot0n committed:
fix: only allow queue_in_background in customize form

phot0n committed:
chore: add confirm dialog for unlock doc button

phot0n committed:
feat: reload listener for form

phot0n committed:
fix: dont repeatedly use get_doc for unlocking reference doc

* chore: add weird behaviour note on unlock_doc method

Commit:
Merge branch 'develop' of https://github.com/Aradhya-Tripathi/frappe into bg-submissions

Commit:
test: Testing for completion

Commit:
feat: Adding exception to submisison queue while unlocking doc

Commit:
refactor: using after commit in enqueuing & removed session time logic

Commit:
test: fixing tests

Commit:
feat: Adding more data to monitor

Commit:
feat: Added execution time to monitor

phot0n commented on frappe/core/doctype/submission_queue/submission_queue.py:
```diff
@@ -0,0 +1,197 @@
+			# Defining job exception when unlocking document.
+			registry = FailedJobRegistry(queue=get_queue(qtype="default"))
+			# If job id is None then this job won't exist in the failed registry
+			for jid in registry.get_job_ids():
+				if jid == job_id:
+					job = Job.fetch(job_id, connection=get_redis_conn())
+					values = {"status": "Failed", "exception": job.exc_info}
```
we're essentially doing job.fetch right? - can;t we just use that instead of searching for it in failed job registry (?)

Aradhya-Tripathi replied:
Was messing around with job exceptions in the registry changed it back now

phot0n commented on frappe/core/doctype/submission_queue/submission_queue.py:
```diff
@@ -0,0 +1,197 @@
+			self.name,
+			{"job_id": current_job.id},
+			update_modified=False,
+		)
```
add a commit after this - rather take out the set_value and commit to a separate method ..something like `update_job_id`

Aradhya-Tripathi replied:
done 👍🏿

phot0n commented on frappe/core/doctype/submission_queue/submission_queue.py:
```diff
@@ -0,0 +1,197 @@
+				{
+					"message": message
+					+ f". View it <a href='/app/{doctype.lower().replace(' ', '-')}/{docname}'><b>here</b></a>"
+					+ f". Execution time {round(float(time_diff), 2)} seconds",
```
this is too much info :sweat_smile: - we don't need to dump so much info upon the user in an alert - and since this will be sent in a 60 sec window of the queue's creation, it doesn't really matter that much :P

Aradhya-Tripathi replied:
ok cool

phot0n commented on frappe/core/doctype/submission_queue/submission_queue.py:
```diff
@@ -0,0 +1,197 @@
+			notify_to = frappe.db.get_value("User", self.enqueued_by, fieldname="email")
+			enqueue_create_notification([notify_to], notification_doc)
+
+	def _unlock_reference_doc(self):
```
this whole method is not working - and gives an "UnboundLocalError" error - since there is no status variable defined in the case `job_id` is found :(

Aradhya-Tripathi replied:
Fixed

Commit:
fix: unlock ref doc

Aradhya-Tripathi committed:
Merge branch 'develop' into bg-submissions

Commit:
refactor: only showing unlock button if job id is defined

Commit:
feat: Added finished condition while unlocking

phot0n committed:
fix: try except for nosuchjoberror in _unlock_reference_doc

* fix: condition for not unlocking reference doc

* chore: change id param -> job_id for update_job_id

phot0n committed:
fix: submission queue banner

if last_failed_submission and last_submissions are equal, show only last_failed_submission

phot0n committed:
chore: remove reload listener from form

phot0n committed:
feat: queue submission for bulk submit

phot0n committed:
fix: use quoted doctype and docname in alert

phot0n committed:
fix: use index for ref_docname in submission queue doctype

Commit:
refactor: checking docstatus before asserting failure

Commit:
refactor: leaving job status as it is on completion

Commit:
refactor: lint

fix: removed time.sleep

phot0n commented:
failing test is unrelated:
![Screenshot from 2022-11-12 20-04-32](https://user-images.githubusercontent.com/32034600/201479062-16a5257a-66cf-4219-b3cb-9d776056c4c7.png)

phot0n commented:
great work @Aradhya-Tripathi :tada:
